### PR TITLE
Rewrite traffic_layout with ArgParser

### PR DESCRIPTION
**Pull request by chitianhao**
This is the first trial using ArgParser. It turns out traffic_layout is super clean and more understandable. It's about 300 lines less than the previous messy code. It's also easier for people to maintain this code in the future.

Note: `runroot_path.yml` is changed to `runroot.yaml`.

### CODE REVIEW

File: src/tscore/runroot.cc [hunk 1]
```diff
@@ -57,9 +56,9 @@ get_yaml_path(const std::string &path, bool json)
 {
   std::string yaml_file;
   if (ink_file_is_directory(path.c_str())) {
-    yaml_file = Layout::relative_to(path, "runroot_path.yml");
+    yaml_file = Layout::relative_to(path, "runroot.yml");
```
Comment (dragon512) [inline src/tscore/runroot.cc]: Can we fix this up so it can look for the existing runroot_path.yml file and the runroot.yml ( and runroot.yaml)

This change is breaking for people using this feature already. Also Trafficserver is using .yaml for all the YAML file we make

File: src/traffic_layout/engine.cc [hunk 2]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
...
+  yamlfile << YAML::BeginMap;
+  for (uint i = 0; i < dir_vector.size(); i++) {
+    yamlfile << YAML::Key << dir_vector[i];
+    yamlfile << YAML::Value << path_map[dir_vector[i]];
+  }
+  yamlfile << YAML::EndMap;
+
+  // create the file
+  std::ofstream f(Layout::relative_to(ts_runroot, "runroot.yaml"));
+  f << yamlfile.c_str();
```
Comment (SolidWallOfCode) [inline src/traffic_layout/engine.cc]: Perhaps check if the file open didn't work and provide an error message?

File: src/traffic_layout/engine.cc [hunk 3]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
...
+}
+
+// check if user want to force remove the ts_runroot
+// return true if user replies Y
+static bool
+check_remove_force()
+{
+  // check for Y/N 3 times
+  for (int i = 0; i < 3; i++) {
+    std::cout << "Are you sure to force removing runroot? (irreversible) Y/N: ";
```
Comment (SolidWallOfCode) [inline src/traffic_layout/engine.cc]: Provide different message here, to make it clear it's repeatedly deliberately and not due to a bug. E.g.
"Are you sure you want to remove runroot?"
"Are  you really sure you want to forcibly remove runroot?"
"I ask you three times, and three times you must agree - remove runroot?"

File: src/traffic_layout/engine.cc [hunk 4]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
...
+}
+
+// the function for removing the runroot
+void
+LayoutEngine::remove_runroot()
+{
+  std::string path = path_handler(arguments.get("path").value(), false, _argv[0]);
+  // check validity of the path
+  if (path.empty()) {
+    ink_fatal("Path not valid (runroot.yaml not found)");
```
Comment (SolidWallOfCode) [inline src/traffic_layout/engine.cc]: Isn't that more "path not provided"?

File: src/traffic_layout/engine.cc [hunk 5]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
...
+    std::cout << "Forcing removing runroot ..." << std::endl;
+    if (!remove_directory(clean_root)) {
+      ink_warning("Failed force removing runroot '%s' - %s", clean_root.c_str(), strerror(errno));
+    }
+  } else {
+    // handle the map and deleting of each directories specified in the yml file
+    RunrootMapType map = runroot_map(Layout::relative_to(clean_root, "runroot.yaml"));
+    map.erase(LAYOUT_PREFIX);
+    map.erase(LAYOUT_EXEC_PREFIX);
+    for (auto it : map) {
```
Comment (SolidWallOfCode) [inline src/traffic_layout/engine.cc]: `for ( auto && i : map) {`

File: src/traffic_layout/engine.cc [hunk 6]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
...
+      std::string dir = it.second;
+      append_slash(dir);
```
Comment (SolidWallOfCode) [inline src/traffic_layout/engine.cc]: OOTPRS, but this would be a nice place to use `ts::file`.

File: src/traffic_layout/engine.cc [hunk 7]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
...
+      std::string later_dir = dir.substr(clean_root.size());
+      dir                   = Layout::relative_to(clean_root, later_dir.substr(0, later_dir.find_first_of("/")));
```
Comment (SolidWallOfCode) [inline src/traffic_layout/engine.cc]: You should be using `string_view` or `TextView` here, not creating substrings with `std::string`.

File: src/traffic_layout/traffic_layout.cc [hunk 8]
```diff
@@ -21,158 +21,67 @@
...
+    .add_option("--path", "", "Specify the path of the runroot", "", 1);
+
+  engine.arguments = engine.parser.parse(argv);
+
+  runroot_handler(argv, engine.arguments.get("json"));
   Layout::create();
 
-  // check the command to execute
-  if (engine.run_flag) {
+  if (engine.arguments.get("init")) {
```
Comment (SolidWallOfCode) [inline src/traffic_layout/traffic_layout.cc]: Why is this here? I thought you could install these functions directly in the argument parser.

File: src/traffic_layout/engine.cc (hunk 4 above)
Comment (chitianhao) [reply src/traffic_layout/engine.cc]: The path_handler returns an empty string if the `runroot.yaml` is not found even if we give provide a path. It should be `not valid`.

File: src/traffic_layout/engine.cc [hunk 9]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
...
+  std::ofstream f(Layout::relative_to(ts_runroot, "runroot.yaml"));
+  f << yamlfile.c_str();
+}
+
+// check if user want to force remove the ts_runroot
+// return true if user replies Y
+static bool
+check_remove_force()
+{
+  // check for Y/N 3 times
```
Comment (bryancall) [inline src/traffic_layout/engine.cc]: I don't think we need to ask for Y 3 times.  Adding --force and asking once is good enough.

File: src/traffic_layout/engine.cc [hunk 10]
```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
   } else {
     path_map[LAYOUT_PREFIX] = ".";
   }
+
+  YAML::Emitter yamlfile;
+  // emit key value pairs to the yaml file
+  yamlfile << YAML::BeginMap;
+  for (uint i = 0; i < dir_vector.size(); i++) {
```
Comment (bryancall) [inline src/traffic_layout/engine.cc]: A ranged based for loop would work well here.

File: src/traffic_layout/engine.cc (hunk 9 above)
Comment (chitianhao) [reply src/traffic_layout/engine.cc]: It's once. The comment is a bit confusing. If the user provides an invalid input it will continue to ask.

**Comment by chitianhao** [comment]
Any other concerns?

File: src/tscore/runroot.cc [hunk 11]
```diff
@@ -53,25 +52,38 @@ static std::string runroot_file = {};
...
   std::ifstream check_file;
+  std::ifstream check_file2;
+  std::ifstream check_file3;
   check_file.open(yaml_file);
-  if (!check_file.good()) {
-    if (!json) {
-      ink_warning("Unable to access runroot: '%s' - %s", yaml_file.c_str(), strerror(errno));
-    }
-    return {};
+  check_file2.open(yaml_file2);
```
Comment (dragon512) [inline src/tscore/runroot.cc]: We should be using the filesystem wrapper we have. ie fs::exists(name) logic .. not trying to open it. We can give better error messages when it exists but has bad permission then just doing an open
